src/compiler/crystal/codegen/fun.cr

```diff
@@ -11,9 +11,7 @@ class Crystal::CodeGenVisitor
 
   def main_fun(name)
     func = @main_mod.functions[name]?
-    unless func
-      raise "BUG: #{name} is not defined"
-    end
+    unreachable! "#{name} is not defined" unless func
```

Comment:
See https://github.com/crystal-lang/crystal/pull/4779#issuecomment-319662762. I think it is related.